File: Data Structures & Algorithms/invert-a-binary-tree/submission-8.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        # Return None if the root is None
        if not root:
            return None

        q = deque()
        q.append(root)

        # Exit the while loop when the queue is empty
        while q:
            # Get the node from the queue to process
            # and remove that node form the queue
            node: TreeNode = q.popleft()

            # Swap the node's left and right children
            node.left, node.right = node.right, node.left

            # Push the swapped left and right children into the queue
            # for next procesing iteration in the while loop
            # Note that we only push existed child into the queue
            # because there's nothing to process if that child node doesn't exist
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        # Return the modified tree's root node
        return root

File: Data Structures & Algorithms/depth-of-binary-tree/submission-8.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    int maxDepth(TreeNode* root) {
        // Return 0 if the root is nullptr
        if (root == nullptr) {
            return 0;
        }

        int result = 0;

        // Create a stack to store a pair of node and depth
        // and push the root node with depth 1 into the stack
        stack<pair<TreeNode*, int>> s;
        s.push({root, 1});

        // Exit the while loop when the stack is empty
        while (!s.empty()) {
            // Get the current node and depth from the stack
            const pair<TreeNode*, int> curr = s.top();
            s.pop();
            const TreeNode* node = curr.first;
            const int depth = curr.second;

            // If that node exists,
            //      1. Push that node's left/right children and incremented depth into the stack
            //      2. Calculate the max depth
            if (node != nullptr) {
                s.push({node->left, depth + 1});
                s.push({node->right, depth + 1});
                result = max(result, depth);
            }
        }
        return result;
    }
};
